### ISSUE TITLE
aria-required-children should fail when children are stolen

### ISSUE THREAD

**Issue by WilcoFiers** [dequelabs/axe-core#926]
This one is fairly minor, but it's wrong. The following should fail aria-required-children:

```html
<div role="listbox" aria-owns="opt2">
  <p role="option">option 1</p>
</div>

<div role="listbox">
  <!-- this element should fail aria-required-children because it's child element is stolen -->
  <p role="option" id="opt2">option 2</p>
</div>
```

I think the best way to solve this would be to add `.getOwned` / `.getOwner` methods. We could add them to the virtual node and memoize them. I'm hoping to do that with other methods too.

**Comment by dylanb** [comment]
Is it documented in ARIA which "parent" takes precedence?

**Comment by WilcoFiers** [comment]
Not really. There's the following definition which sort of implies it that it's exclusive:

> **owning element**
> An 'owning element' is any DOM ancestor of the element, or any element with an aria-owns attribute which references the ID of the element.
>
> Source: https://www.w3.org/TR/html-aam-1.0/#dfn-owning-element

I certainly couldn't find anything definitive. I found this while testing with VO.

**Comment by WilcoFiers** [comment]
We should do screen reader testing before we build on this.

**Comment by scurker** [comment]
I did some [initial testing](https://codepen.io/scurker/pen/XwwjMB) on the Mac side to see how the following was being read:

```html
<div role="listbox" aria-owns="opt2">
  <p role="option">Inherited Child</p>
<div>
<div role="listbox">
  <p role="option" id="opt2">Stolen Child</p>
<div>
```

| | VoiceOver on Chrome 74 & MacOS Mojave | VoiceOver on Edge 76 & MacOS Mojave | VoiceOver on Safari 12.03 & MacOS Mojave | ChromeVox on Chrome 74 & MacOS Mojave | VoiceOver on Safari & iOS 12 |
| - | --- | --- | --- | --- | --- |
| Listbox 1 | Inherited Child, 1 of 2 | Inherited Child, 1 of 2 | Inherited Child, Text, 1 of 1 | Inherited Child, 1 of 1, Listbox | Inherited Child, List Start |
| Listbox 2 | Empty | Empty | Stolen Child, Text, 1 of 1 | Stolen Child, 1 of 1, Listbox | Stolen Child, List Start |

So far, only Chromium + VoiceOver is announcing the first listbox with the stolen child. All other combinations are announcing each list box with their direct descendants. I'll be curious to see how JAWS handles this on the Windows side, but there doesn't appear to be parity across readers on the Mac side.

**Comment by straker** [comment]
I tested JAWS 2019.1903.47 and IE on Windows and it didn't read out the number of items in the list, so it just read each option individually.

> Inherited Child, listbox item
> Stolen Child, listbox item


NVDA 2019.1 and Firefox 65.0.2 on Windows reads the stolen child and thinks the second one is empty

> List Inherited Child
> List

**Comment by dylanb** [comment]
So the bottom line is that this is not accessibility supported at this point in time, however we could still implement this as a best practice rule and then switch it to a violation once the support catches up

**Comment by WilcoFiers** [comment]
Talked to @dylanb, he agreed that this is a WCAG issue not a best practice. We'll have to come up with a fix for it before we proceed.

**Comment by straker** [comment]
So I'm not sure this should be part of `aria-required-children`. Required children only looks at the existence of certain elements in the tree. If we throw an error for children been stolen by a different tree, the error message for _why_ `aria-required-children` won't make sense 

> Required ARIA child role not present: option

We should probably create an new rule that looks for this and gives a more helpful error message that describes that the child does not belong to the tree anymore because another element owns it, and probably point to the other element so they can more easily find it.

**Comment by WilcoFiers** [comment]
I think this should probably be a new "none" check for aria-required-children, that then reports this as incomplete. I'm going to keep this open, this seems like something we should do.
